File: packages/boxsh/src/shell.ts
// Shell parsing and execution.
import type { StorageBackend } from "./backend.js";
import { normalize } from "./backend.js";
import type { Engine } from "./engine.js";

const enc = new TextEncoder();
const dec = new TextDecoder();

export interface ShellSession {
  env: Record<string, string>;
  cwd: string;
  lastStatus: number;
}

export interface ScriptResult {
  stdout: Uint8Array;
  stderr: Uint8Array;
  code: number;
}

interface Token {
  word?: string;
  op?: string;
}

interface Stage {
  argv: string[];
  in: string | null;
  out: string | null;
  append: boolean;
}

interface Chain {
  pipeline: Stage[];
  joiner: string | null;
}

const OPS = ["&&", "||", ";", "|", ">>", ">", "<"];
const HEREDOC_RE = /<<-?\s*(?:'([A-Za-z_]\w*)'|"([A-Za-z_]\w*)"|([A-Za-z_]\w*))/;

const concat = (chunks: Uint8Array[]): Uint8Array => {
  const out = new Uint8Array(chunks.reduce((n, c) => n + c.length, 0));
  let at = 0;
  for (const c of chunks) {
    out.set(c, at);
    at += c.length;
  }
  return out;
};

export function createShell(
  engine: Engine,
  backendRef: { current: StorageBackend },
  session: ShellSession,
) {
  const expand = (s: string): string =>
    s.replace(/\$(\?|\{[A-Za-z_][A-Za-z0-9_]*\}|[A-Za-z_][A-Za-z0-9_]*)/g, (_, k: string) => {
      const key = k === "?" ? "?" : k.replace(/[{}]/g, "");
      return key === "?" ? String(session.lastStatus) : (session.env[key] ?? "");
    });

  let subDepth = 0;
  function expandSubs(line: string): string {
    if (!line.includes("$(")) return line;
    if (subDepth > 8) throw new Error("command substitution nested too deep");
    let out = "";
    let i = 0;
    let sq = false;
    while (i < line.length) {
      const c = line[i];
      if (c === "'") {
        sq = !sq;
        out += c;
        i++;
      } else if (!sq && c === "$" && line[i + 1] === "(") {
        let depth = 1;
        let j = i + 2;
        for (; j < line.length && depth; j++) {
          if (line[j] === "(") depth++;
          else if (line[j] === ")") depth--;
        }
        if (depth) throw new Error("unterminated $( )");
        subDepth++;
        let text: string;
        try {
          text = dec.decode(execLine(line.slice(i + 2, j - 1), null, true).captured);
        } finally {
          subDepth--;
        }
        out += text.replace(/\s+/g, " ").trim();
        i = j;
      } else {
        out += c;
        i++;
      }
    }
    return out;
  }

  function tokenize(line: string): Token[] {
    const tokens: Token[] = [];
    let i = 0;
    let cur = "";
    let started = false;
    const flush = () => {
      if (started) tokens.push({ word: cur });
      cur = "";
      started = false;
    };
    while (i < line.length) {
      const c = line[i];
      if (c === "'") {
        started = true;
        const end = line.indexOf("'", i + 1);
        if (end === -1) throw new Error("unterminated '");
        cur += line.slice(i + 1, end);
        i = end + 1;
      } else if (c === '"') {
        started = true;
        let j = i + 1;
        let seg = "";
        for (; j < line.length && line[j] !== '"'; j++) {
          if (line[j] === "\\" && '"$\\'.includes(line[j + 1])) seg += line[++j];
          else seg += line[j];
        }
        if (j >= line.length) throw new Error('unterminated "');
        cur += expand(seg);
        i = j + 1;
      } else if (c === "\\") {
        started = true;
        cur += line[i + 1] ?? "";
        i += 2;
      } else if (c === " " || c === "\t") {
        flush();
        i++;
      } else {
        const op = OPS.find((o) => line.startsWith(o, i));
        if (op) {
          flush();
          tokens.push({ op });
          i += op.length;
        } else if (c === "$") {
          const m = /^\$(\?|\{[A-Za-z_][A-Za-z0-9_]*\}|[A-Za-z_][A-Za-z0-9_]*)/.exec(line.slice(i));
          if (m) {
            started = true;
            const key = m[1] === "?" ? "?" : m[1].replace(/[{}]/g, "");
            cur += key === "?" ? String(session.lastStatus) : (session.env[key] ?? "");
            i += m[0].length;
          } else {
            started = true;
            cur += c;
            i++;
          }
        } else {
          started = true;
          cur += c;
          i++;
        }
      }
    }
    flush();
    return tokens;
  }

  function parse(tokens: Token[]): Chain[] {
    const chains: Chain[] = [];
    let stages: Stage[] = [];
    let stage: Stage = { argv: [], in: null, out: null, append: false };
    let expect: "in" | "out" | null = null;
    const endStage = () => {
      stages.push(stage);
      stage = { argv: [], in: null, out: null, append: false };
    };
    const endChain = (joiner: string | null) => {
      endStage();
      chains.push({ pipeline: stages, joiner });
      stages = [];
    };
    for (const t of tokens) {
      if (t.word !== undefined) {
        if (expect === "in") {
          stage.in = t.word;
          expect = null;
        } else if (expect === "out") {
          stage.out = t.word;
          expect = null;
        } else stage.argv.push(t.word);
      } else if (t.op === "<") expect = "in";
      else if (t.op === ">") {
        stage.append = false;
        expect = "out";
      } else if (t.op === ">>") {
        stage.append = true;
        expect = "out";
      } else if (t.op === "|") endStage();
      else endChain(t.op ?? null);
    }
    endChain(null);
    return chains.filter((c) => c.pipeline.some((s) => s.argv.length));
  }

  type BuiltinResult = { out?: string; err?: string; code: number };
  const builtins: Record<string, (args: string[]) => BuiltinResult> = {
    ":": () => ({ code: 0 }),
    cd(argv) {
      const target = normalize(
        (argv[0] ?? session.env.HOME ?? "/").startsWith("/")
          ? (argv[0] ?? session.env.HOME ?? "/")
          : `${session.cwd}/${argv[0]}`,
      );
      const e = backendRef.current.entry(target);
      if (!e) return { err: `cd: ${argv[0] ?? ""}: No such file or directory\n`, code: 1 };
      if (e.kind !== "dir") return { err: `cd: ${argv[0]}: Not a directory\n`, code: 1 };
      session.cwd = "/" + target;
      return { code: 0 };
    },
    pwd: () => ({ out: session.cwd + "\n", code: 0 }),
    export(argv) {
      for (const a of argv) {
        const eq = a.indexOf("=");
        if (eq > 0) session.env[a.slice(0, eq)] = a.slice(eq + 1);
      }
      return { code: 0 };
    },
    unset(argv) {
      for (const a of argv) delete session.env[a];
      return { code: 0 };
    },
    env: () => ({
      out:
        Object.entries({ ...session.env, PWD: session.cwd })
          .map(([k, v]) => `${k}=${v}`)
          .join("\n") + "\n",
      code: 0,
    }),
  };

  const resolvePath = (p: string): string =>
    normalize(p.startsWith("/") ? p : `${session.cwd}/${p}`);

  interface LineResult {
    stdout: Uint8Array[];
    stderr: Uint8Array[];
    captured: Uint8Array;
  }

  function execLine(
    line: string,
    heredocBody: { raw: string; quoted: boolean } | null = null,
    capture = false,
  ): LineResult {
    const stdout: Uint8Array[] = [];
    const stderr: Uint8Array[] = [];
    const captured: Uint8Array[] = [];
    const done = (): LineResult => ({ stdout, stderr, captured: concat(captured) });
    const fail = (msg: string): LineResult => {
      stderr.push(enc.encode(`boxsh: ${msg}\n`));
      session.lastStatus = 2;
      return done();
    };

    const fm = /^\s*for\s+([A-Za-z_]\w*)\s+in\s+(.+?)\s*;\s*do\s+(.+?)\s*;?\s*done\s*$/.exec(line);
    if (fm) {
      let words: string[];
      try {
        words = expand(expandSubs(fm[2])).split(/\s+/).filter(Boolean);
      } catch (e) {
        return fail((e as Error).message);
      }
      for (const w of words) {
        session.env[fm[1]] = w;
        const r = execLine(fm[3], null, capture);
        stdout.push(...r.stdout);
        stderr.push(...r.stderr);
        if (capture && r.captured.length) captured.push(r.captured);
      }
      return done();
    }

    let chains: Chain[];
    try {
      chains = parse(tokenize(expandSubs(line)));
    } catch (e) {
      return fail((e as Error).message);
    }

    let body = heredocBody;
    let skip: string | null = null;
    for (const { pipeline, joiner } of chains) {
      if (skip === "&&" && session.lastStatus !== 0) {
        skip = joiner;
        continue;
      }
      if (skip === "||" && session.lastStatus === 0) {
        skip = joiner;
        continue;
      }
      let data: Uint8Array = body
        ? enc.encode(body.quoted ? body.raw : expand(body.raw))
        : new Uint8Array(0);
      for (let s = 0; s < pipeline.length; s++) {
        const st = pipeline[s];
        const name = st.argv[0];
        if (st.in) {
          const e = backendRef.current.read(resolvePath(st.in));
          if (e === undefined) {
            stderr.push(enc.encode(`boxsh: ${st.in}: No such file or directory\n`));
            session.lastStatus = 1;
            break;
          }
          data = e;
        }
        let r: { out: Uint8Array; err: Uint8Array; code: number };
        if (builtins[name]) {
          const b = builtins[name](st.argv.slice(1));
          r = { out: enc.encode(b.out ?? ""), err: enc.encode(b.err ?? ""), code: b.code };
        } else if (engine.knows(name)) {
          r = engine.run(st.argv, data);
        } else {
          r = {
            out: new Uint8Array(0),
            err: enc.encode(`boxsh: ${name}: command not found\n`),
            code: 127,
          };
        }
        if (r.err.length) stderr.push(r.err);
        session.lastStatus = r.code;
        data = r.out;
        if (s === pipeline.length - 1) {
          if (st.out) {
            try {
              backendRef.current.write(resolvePath(st.out), data);
            } catch (e) {
              stderr.push(enc.encode(`boxsh: ${st.out}: ${(e as Error).message}\n`));
              session.lastStatus = 1;
            }
          } else if (capture) {
            if (data.length) captured.push(data);
          } else if (data.length) stdout.push(data);
        }
      }
      skip = joiner;
    }
    return done();
  }

  /** Execute a script: one or more lines, heredocs included. */
  function execScript(script: string): ScriptResult {
    const stdout: Uint8Array[] = [];
    const stderr: Uint8Array[] = [];
    const lines = script.split("\n");
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (!line.trim()) continue;
      const m = HEREDOC_RE.exec(line);
      let body: { raw: string; quoted: boolean } | null = null;
      let cmd = line;
      if (m) {
        const tag = m[1] ?? m[2] ?? m[3];
        const quoted = Boolean(m[1] ?? m[2]);
        cmd = line.replace(m[0], "");
        const collected: string[] = [];
        i++;
        while (i < lines.length && lines[i] !== tag) collected.push(lines[i++]);
        body = { raw: collected.map((l) => l + "\n").join(""), quoted };
      }
      const r = execLine(cmd, body);
      stdout.push(...r.stdout);
      stderr.push(...r.stderr);
    }
    return { stdout: concat(stdout), stderr: concat(stderr), code: session.lastStatus };
  }

  return { execScript };
}

File: crates/boxsh-shell/tests/shell.rs
//! Shell language semantics, pinned against the TypeScript shell this crate
//! ports (shell.ts + the shell sections of packages/boxsh/test/api.mjs).
//! Commands run through a scripted double — the shell under test is real,
//! the coreutils are not (they have their own crates and tests).

use boxsh_fs::{Backend, MemoryBackend};
use boxsh_shell::{CommandOutput, CommandRunner, Env, ScriptResult, Session, exec_script};

/// Minimal stdin/argv-pure stand-ins for the commands api.mjs exercises.
struct Utils;

impl CommandRunner for Utils {
    fn knows(&self, name: &str) -> bool {
        matches!(
            name,
            "echo" | "cat" | "seq" | "tail" | "sort" | "wc" | "grep" | "tr" | "true" | "false"
        )
    }

    fn run(&mut self, argv: &[String], stdin: &[u8], _session: &Session) -> CommandOutput {
        let ok = |out: String| CommandOutput {
            out: out.into_bytes(),
            err: Vec::new(),
            code: 0,
        };
        let text = || String::from_utf8_lossy(stdin).to_string();
        let lines = |s: &str| s.lines().map(str::to_string).collect::<Vec<_>>();
        match argv[0].as_str() {
            "true" => ok(String::new()),
            "false" => CommandOutput {
                out: Vec::new(),
                err: Vec::new(),
                code: 1,
            },
            "echo" => ok(format!("{}\n", argv[1..].join(" "))),
            "cat" => CommandOutput {
                out: stdin.to_vec(),
                err: Vec::new(),
                code: 0,
            },
            "seq" => {
                let a: i64 = argv[1].parse().unwrap();
                let b: i64 = argv[2].parse().unwrap();
                ok((a..=b).map(|n| format!("{n}\n")).collect())
            }
            "tail" => {
                let n: usize = argv[1].trim_start_matches('-').parse().unwrap();
                let all = text();
                let l = lines(&all);
                ok(l[l.len().saturating_sub(n)..]
                    .iter()
                    .map(|s| format!("{s}\n"))
                    .collect())
            }
            "sort" => {
                let all = text();
                let mut l = lines(&all);
                l.sort();
                if argv.get(1).map(String::as_str) == Some("-r") {
                    l.reverse();
                }
                ok(l.iter().map(|s| format!("{s}\n")).collect())
            }
            "wc" => ok(format!(
                "{}\n",
                stdin.iter().filter(|&&b| b == b'\n').count()
            )),
            "grep" => {
                let (count, pat) = if argv[1] == "-c" {
                    (true, argv[2].as_str())
                } else {
                    (false, argv[1].as_str())
                };
                let all = text();
                let hits: Vec<&str> = all.lines().filter(|l| l.contains(pat)).collect();
                let out = if count {
                    format!("{}\n", hits.len())
                } else {
                    hits.iter().map(|l| format!("{l}\n")).collect()
                };
                let code = if hits.is_empty() { 1 } else { 0 };
                CommandOutput {
                    out: out.into_bytes(),
                    err: Vec::new(),
                    code,
                }
            }
            "tr" => ok(text().to_uppercase()),
            other => panic!("double has no {other}"),
        }
    }
}

struct Harness {
    backend: MemoryBackend,
    session: Session,
}

impl Harness {
    fn new() -> Self {
        let mut session = Session::new();
        session.env = Env::from_iter([("HOME", "/"), ("USER", "agent")]);
        Harness {
            backend: MemoryBackend::new(),
            session,
        }
    }

    fn exec(&mut self, script: &str) -> ScriptResult {
        exec_script(&mut self.backend, &mut Utils, &mut self.session, script)
    }

    fn out(&mut self, script: &str) -> String {
        let r = self.exec(script);
        String::from_utf8(r.stdout).unwrap()
    }

    fn err(&mut self, script: &str) -> (String, i32) {
        let r = self.exec(script);
        (String::from_utf8(r.stderr).unwrap(), r.code)
    }
}

#[test]
fn echo_env_and_state_persist_across_execs() {
    let mut h = Harness::new();
    assert_eq!(h.out("echo hello from $USER"), "hello from agent\n");
    h.backend.mkdir("src").unwrap();
    h.exec("export GREETING=hey && cd /src");
    assert_eq!(h.session.cwd, "/src");
    assert_eq!(h.out("echo $GREETING $(pwd)"), "hey /src\n");
}

#[test]
fn pipelines_redirects_and_relative_paths() {
    let mut h = Harness::new();
    h.backend.mkdir("src").unwrap();
    h.exec("cd /src");
    h.exec("seq 1 100 | tail -3 | sort -r > top.txt");
    assert_eq!(h.backend.read("src/top.txt").unwrap(), b"99\n98\n100\n");
}

#[test]
fn heredoc_multiline_script() {
    let mut h = Harness::new();
    let r = h.out("cat <<EOF > note.txt\nline one\nline two\nEOF\nwc -l < note.txt");
    assert_eq!(r.trim(), "2");
    assert_eq!(h.backend.read("note.txt").unwrap(), b"line one\nline two\n");
}

#[test]
fn heredoc_quoting_controls_expansion() {
    let mut h = Harness::new();
    h.exec("export NAME=world");
    assert_eq!(h.out("cat <<EOF\nhi $NAME\nEOF"), "hi world\n");
    assert_eq!(h.out("cat <<'EOF'\nhi $NAME\nEOF"), "hi $NAME\n");
    assert_eq!(h.out("cat <<\"EOF\"\nhi $NAME\nEOF"), "hi $NAME\n");
}

#[test]
fn command_not_found_is_a_code_not_a_crash() {
    let mut h = Harness::new();
    let (err, code) = h.err("definitely-not-a-command");
    assert_eq!(code, 127);
    assert!(err.contains("command not found"), "{err}");
}

#[test]
fn quoting_rules() {
    let mut h = Harness::new();
    h.exec("export V=val");
    assert_eq!(h.out("echo '$V literal'"), "$V literal\n");
    assert_eq!(h.out("echo \"$V quoted\""), "val quoted\n");
    assert_eq!(h.out("echo \"a\\\"b \\$V \\\\\""), "a\"b $V \\\n");
    assert_eq!(h.out("echo a\\ b"), "a b\n");
    assert_eq!(h.out("echo ''"), "\n"); // empty word survives
    let (err, code) = h.err("echo 'unterminated");
    assert_eq!(code, 2);
    assert!(err.contains("unterminated '"));
    let (err, code) = h.err("echo \"unterminated");
    assert_eq!(code, 2);
    assert!(err.contains("unterminated \""));
}

#[test]
fn variable_forms() {
    let mut h = Harness::new();
    h.exec("export AB=x ABC=y");
    assert_eq!(h.out("echo $AB-$ABC ${AB}C"), "x-y xC\n");
    assert_eq!(h.out("echo $MISSING."), ".\n");
    assert_eq!(h.out("echo $ $5"), "$ $5\n"); // no ident: literal
    h.exec("false");
    assert_eq!(h.out("echo $?"), "1\n");
    h.exec("true");
    assert_eq!(h.out("echo $?"), "0\n");
}

#[test]
fn command_substitution() {
    let mut h = Harness::new();
    assert_eq!(h.out("echo [$(echo inner)]"), "[inner]\n");
    // Captured output collapses internal whitespace.
    assert_eq!(h.out("echo $(seq 1 3)"), "1 2 3\n");
    // Nesting.
    assert_eq!(h.out("echo $(echo $(echo deep))"), "deep\n");
    // Single quotes suppress it.
    assert_eq!(h.out("echo '$(echo nope)'"), "$(echo nope)\n");
    let (err, code) = h.err("echo $(echo unterminated");
    assert_eq!(code, 2);
    assert!(err.contains("unterminated $( )"));
}

#[test]
fn substitution_depth_is_capped() {
    let mut h = Harness::new();
    let mut script = "echo hi".to_string();
    for _ in 0..12 {
        script = format!("echo $({script})");
    }
    // The innermost level fails and the message bubbles up through the
    // capture chain; the outer echoes still run (with empty output).
    let (err, _) = h.err(&script);
    assert!(err.contains("nested too deep"), "{err}");
}

#[test]
fn substitution_stderr_reaches_the_caller() {
    let mut h = Harness::new();
    let r = h.exec("echo [$(no-such-cmd)]");
    assert_eq!(r.stdout, b"[]\n");
    assert!(
        String::from_utf8(r.stderr)
            .unwrap()
            .contains("no-such-cmd: command not found")
    );
}

#[test]
fn conditional_chains() {
    let mut h = Harness::new();
    assert_eq!(h.out("true && echo yes"), "yes\n");
    assert_eq!(h.out("false && echo yes"), "");
    assert_eq!(h.out("false || echo fallback"), "fallback\n");
    assert_eq!(h.out("true || echo fallback"), "");
    assert_eq!(h.out("false ; echo always"), "always\n");
    // Skip chains through the joiner that follows the skipped chain.
    assert_eq!(h.out("false && echo a || echo b"), "b\n");
    assert_eq!(h.out("true && echo a || echo b"), "a\n");
}

#[test]
fn redirects() {
    let mut h = Harness::new();
    h.exec("echo one > f.txt");
    assert_eq!(h.backend.read("f.txt").unwrap(), b"one\n");
    h.exec("echo two > f.txt");
    assert_eq!(h.backend.read("f.txt").unwrap(), b"two\n");
    h.exec("echo three >> f.txt");
    assert_eq!(h.backend.read("f.txt").unwrap(), b"two\nthree\n");
    h.exec("echo fresh >> new.txt"); // append to a missing file creates it
    assert_eq!(h.backend.read("new.txt").unwrap(), b"fresh\n");
    assert_eq!(h.out("cat < f.txt"), "two\nthree\n");
    let (err, code) = h.err("cat < missing.txt");
    assert_eq!(code, 1);
    assert!(err.contains("missing.txt: No such file or directory"));
    // Write failure surfaces on stderr with status 1.
    h.backend.mkdir("d").unwrap();
    let (err, code) = h.err("echo x > d");
    assert_eq!(code, 1);
    assert!(err.contains("boxsh: d:"), "{err}");
}

#[test]
fn for_loops() {
    let mut h = Harness::new();
    assert_eq!(h.out("for i in 1 2 3; do echo n$i; done"), "n1\nn2\nn3\n");
    // Words expand, including substitutions; body sees pipes.
    assert_eq!(
        h.out("for w in $(seq 1 2) three; do echo $w | cat; done"),
        "1\n2\nthree\n"
    );
    // Optional trailing semicolon before done.
    assert_eq!(
        h.out("for i in a; do echo $i done marker; done"),
        "a done marker\n"
    );
    // Loop variable persists after the loop, like the original.
    h.exec("for i in last; do true; done");
    assert_eq!(h.out("echo $i"), "last\n");
}

#[test]
fn builtins() {
    let mut h = Harness::new();
    h.backend.mkdir("dir").unwrap();
    h.backend.write("file", b"x").unwrap();

    assert_eq!(h.out("pwd"), "/\n");
    h.exec("cd dir");
    assert_eq!(h.session.cwd, "/dir");
    assert_eq!(h.out("pwd"), "/dir\n");
    h.exec("cd");
    assert_eq!(h.session.cwd, "/"); // HOME
    let (err, code) = h.err("cd missing");
    assert_eq!(code, 1);
    assert!(err.contains("cd: missing: No such file or directory"));
    let (err, code) = h.err("cd file");
    assert_eq!(code, 1);
    assert!(err.contains("cd: file: Not a directory"));

    h.exec("export A=1 B=2 not-an-assignment");
    assert_eq!(h.session.env.get("A"), Some("1"));
    assert_eq!(h.session.env.get("B"), Some("2"));
    h.exec("unset A");
    assert_eq!(h.session.env.get("A"), None);

    // env prints insertion order with PWD appended last.
    let env_out = h.out("env");
    let last = env_out.lines().last().unwrap();
    assert_eq!(last, "PWD=/");
    assert!(env_out.starts_with("HOME=/\nUSER=agent\n"), "{env_out}");

    let r = h.exec(":");
    assert_eq!(r.code, 0);
}

#[test]
fn blank_lines_and_multi_line_scripts() {
    let mut h = Harness::new();
    assert_eq!(h.out("echo a\n\n   \necho b"), "a\nb\n");
    let r = h.exec("false\n");
    assert_eq!(r.code, 1); // script code is the last status
}

#[test]
fn two_sessions_share_one_backend() {
    let mut backend = MemoryBackend::new();
    let mut s1 = Session::new();
    let mut s2 = Session::new();
    exec_script(&mut backend, &mut Utils, &mut s1, "echo shared > f.txt");
    let r = exec_script(&mut backend, &mut Utils, &mut s2, "cat < f.txt");
    assert_eq!(r.stdout, b"shared\n");
}

#[test]
fn empty_pipeline_stage_is_command_not_found() {
    let mut h = Harness::new();
    let (err, code) = h.err("echo a | ");
    assert_eq!(code, 127);
    assert!(err.contains("command not found"));
}
